toolkit/tools/pkg/imagecustomizerlib/defaultgrubutils.go

```diff
@@ -0,0 +1,256 @@
+				}
+
+				// There is a bug in Azure Linux 2.0 and early builds of Azure Linux 3.0 where "$kernelopts" is not
+				// properly escaped.
+				// So, be nice and fix it up.
+				argStringBuilder.WriteString("$")
+				argStringBuilder.WriteString(grubKernelOpts)
+
+			case grub.VAR_EXPANSION:
+				return nil, fmt.Errorf("unexpected variable expansion (%s)", subword.Value)
+			}
+		}
+
+		argValue := argStringBuilder.String()
+		name, value, foundEqSymbol := strings.Cut(argValue, "=")
+		if !foundEqSymbol {
+			// Not a variable assignment.
+			continue
+		}
+
+		varAssign := grubDefaultVarAssign{
+			Token: argToken,
+			Name:  name,
+			Value: value,
+		}
+		varAssigns = append(varAssigns, varAssign)
+	}
+
+	return varAssigns, nil
+}
+
+func findGrubDefaultVarAssign(varAssigns []grubDefaultVarAssign, name string) (grubDefaultVarAssign, error) {
+	for _, varAssign := range varAssigns {
+		if varAssign.Name == name {
+			return varAssign, nil
+		}
+	}
+
+	err := fmt.Errorf("failed to find %s variable assignment (%s)", installutils.GrubDefFile, name)
+	return grubDefaultVarAssign{}, err
+}
+
+func getGrubDefaultLinuxArgs(defaultGrubContent string, varName string,
+) (grubDefaultVarAssign, []grubConfigLinuxArg, int, error) {
+	varAssigns, err := findGrubDefaultVarAssigns(defaultGrubContent)
+	if err != nil {
+		err = fmt.Errorf("failed to parse %s file:\n%w", installutils.GrubDefFile, err)
+		return grubDefaultVarAssign{}, nil, 0, err
+	}
+
+	// Find the variable's (e.g. GRUB_CMDLINE_LINUX) line.
+	cmdLineVarAssign, err := findGrubDefaultVarAssign(varAssigns, varName)
+	if err != nil {
+		return grubDefaultVarAssign{}, nil, 0, err
+	}
+
+	// The (parsed) string value of the variable is copied verbatim into the grub.cfg file.
+	// So, args can have quotes if needed, but those quotes will be double escaped.
+	argsString := cmdLineVarAssign.Value
+	grubTokens, err := grub.TokenizeConfig(argsString)
+	if err != nil {
+		err = fmt.Errorf("failed to parse %s's value:\n%w", varName, err)
+		return grubDefaultVarAssign{}, nil, 0, err
+	}
+
+	var insertAt int
+	if varName == defaultGrubCmdlineLinuxDefault {
+		// GRUB_CMDLINE_LINUX_DEFAULT variable has the $kernelopts arg.
+		// Any args inserted should be inserted before $kernelopts.
+		insertAt, err = findCommandLineInsertAt(grubTokens)
+		if err != nil {
+			err = fmt.Errorf("failed to parse %s's value args:\n%w", varName, err)
+			return grubDefaultVarAssign{}, nil, 0, err
+		}
+	} else {
+		// Insert args at the end of the string.
+		insertAt = len(argsString)
+	}
+
+	args, err := parseCommandLineArgs(grubTokens)
+	if err != nil {
+		err = fmt.Errorf("failed to parse %s's value args:\n%w", varName, err)
+		return grubDefaultVarAssign{}, nil, 0, err
+	}
+
+	return cmdLineVarAssign, args, insertAt, nil
+}
+
+func addExtraCommandLineToDefaultGrubFile(defaultGrubContent string, extraCommandLine string) (string, error) {
```

Comment:
would it make sense to group functions that manipulate the default grub file in a type by itself? The naming is a bit tricky across various functions given that the default can mean the default grub file, or a default property/value. It might be a bit clearer that I use `defaultGrub.DoSomething()` `grubCfg.DoSomething()` to distinguish which file I'm targeting. Or, abstract that for the user so they do not need to know the difference (if possible)...[I need to read about who the files interact]

Reply:
The `BootCustomizer` type is intended to be the abstraction point between the two files. Though, it doesn't cover all scenarios yet.

I changed the naming so that it is "defaultGrubFile" instead of only "defaultGrub".